### DISCUSSION TITLE
Godot Meson port discussion

### DISCUSSION THREAD

**Discussion by jpakkane** [jpakkane/godot#3] [Announcements]
## 👋 Welcome!

This area is for discussions for porting the build system of Godot to Meson.

**Comment by Xrayez** [comment]
I find it essential being able to [compile Godot with external custom modules](https://docs.godotengine.org/en/stable/development/compiling/introduction_to_the_buildsystem.html#custom-modules), the support for which I contributed myself in godotengine/godot#36922, which allows to build modules located from anywhere on the filesystem. There are numerous [Godot modules](https://github.com/topics/godot-module) that can be compiled using existing buildsystem.

I still think that SCons works fine given Godot's design in general. There are also several PRs by @dmoody256 (SCons contributor) https://github.com/godotengine/godot/pulls/dmoody256 that aim to improve the existing buildsystem, which unfortunately haven't received much attention yet, either.

If you really want to increase the chances of integrating Meson in Godot, please think about the use case above, thanks.

**Reply by dmoody256** [reply]
I admit I did lose motivation to continue work on those PRs because I felt the overall developer sentiment seemed to be to get rid of SCons. I still am interested in working with Godot SCons build system if there is any desire to keep SCons build system.

**Reply by Xrayez** [reply]
Yeah, I think the current indecision in Godot is what holds back the work, both for improving SCons usage and possible Meson integration... I think major buildsystem changes are unlikely to be considered up until Godot 4.0 stable is out, but that's just my estimation. 🙂

The [release policy is not really fixed in Godot](https://docs.godotengine.org/en/stable/about/release_policy.html) so even major changes may be part of minor releases, especially when the compatibility changes mostly pertain to the editor, which is seen as the final product over everything else.

I hope SCons remains, especially when it's been used for quite a while in Godot. With things like [Vulkan and lack of GLES](https://godotengine.org/article/about-godot4-vulkan-gles3-and-gles2), [GDScript 2.0](https://godotengine.org/article/gdscript-progress-report-feature-complete-40), [GDExtension](https://godotengine.org/article/introducing-gd-extensions), and possibly Meson, I'm afraid I'd have to keep using Godot 3.x for quite some time. I've also seen some hobbyist Godot 3.x forks popping up because of that.

**Comment by jpakkane** [comment]
> I find it essential being able to [compile Godot with external custom modules](https://docs.godotengine.org/en/stable/development/compiling/introduction_to_the_buildsystem.html#custom-modules), the support for which I contributed myself in https://github.com/godotengine/godot/pull/36922, which allows to build modules located from anywhere on the filesystem.

The current module implementation uses the existing `config.py` things and is very tied to how the current thing is set up. One of my goals is to rewrite that to use Meson builtins and once that is done, building external modules becomes simple. It should be noted, though, that Meson is quite strict about source layout so that the modules must be within a single project, not anywhere in the file system. You can still do Git submodules or whatever you want to join the two.

**Comment by jpakkane** [comment]
I [wrote example code and a blog post](https://nibblestew.blogspot.com/2022/02/supporting-external-modules-in-godot.html) on how external modules with Meson would work.

**Comment by rtiemann** [comment]
Nice to see this Meson port, I hope it will be accepted when done!

I was able to build Godot on Linux, but the result lacks the "opengl3" and "opengl3_es" x11 drivers.


Version built by Meson:

    $ build/godot --help | grep display-driver
    --display-driver <driver>         Display driver (and rendering driver) ['x11' ('vulkan'), 'headless' ('dummy')].

Version built by SCons:

    $ bin/godot.linuxbsd.editor.x86_64 --help | grep display-driver
    --display-driver <driver>         Display driver (and rendering driver) ['x11' ('vulkan', 'opengl3', 'opengl3_es'), 'headless' ('dummy')].

The version without OpenGL3 support shows a popup when trying to load a project: "Your video card drivers seem not to support the required OpenGL 3.3 version."

Output on the console:

    Godot Engine v4.0.dev.custom_build.ACDCABBAC - https://godotengine.org
    Vulkan API 1.2.175 - Forward+ - Using Vulkan Device #0: NVIDIA - NVIDIA GeForce GT 730M

    ERROR: Parameter "renderer_button_group->get_pressed_button()" is null.
    at: _renderer_selected (../editor/project_manager.cpp:400)
    Editing project: /home/rt/coding/src/AndroidBug
    Godot Engine v4.0.dev.custom_build.ACDCABBAC - https://godotengine.org
    ERROR: Video driver not found
    at: DisplayServerX11 (../platform/linuxbsd/x11/display_server_x11.cpp:6098)

As far as I can tell, `platform/linuxbsd/detect.py` defines `GLES3_ENABLED` to enable the video driver if `opengl3` is defined somewhere (I guess this is a configuration option which is enabled by default), but there is no such define in any `meson.build` file. So I tried adding it to the `CPP_ARGS` array in `platform/linuxbsd/meson.build` as `-DGLES3_ENABLED`, which changes the build, but also results in an error:

    In file included from ../platform/linuxbsd/x11/detect_prime_x11.cpp:39:
    ../thirdparty/glad/glad/glx.h:42:10: fatal error: glad/gl.h: No such file or directory
    42 | #include <glad/gl.h>

Seems like this is not supported by the port yet. I guess a `thirdparty/glad/meson.build` file should be added now to tell Meson how to handle `glad`. Am I on the right track or should this work completely differently?

**Reply by jpakkane** [reply]
FWICT that is how it should work. I don't know the exact circumstances under which glad should be enabled (especially on platforms other than Linux), though, but patches welcome and all that.

**Reply by rtiemann** [reply]
> FWICT that is how it should work. I don't know the exact circumstances under which glad should be enabled (especially on platforms other than Linux), though,

OK, I thought that maybe I've missed some configuration switch that needs to be flipped or something. I'll figure out what the SCons code is doing and where the `opengl3` bit is coming from.

> but patches welcome and all that.

Sure, I'll see what I can do. :)

**Comment by Jimmy2099** [comment]
the latest version can not run in window with mscv build


`C:\Users\Administrator\Desktop\ms\godot\complie\godot.exe -e --path C:/Users/Administrator/Documents/11
Godot Engine v4.0.dev.custom_build.ACDCABBAC - https://godotengine.org
Vulkan API 1.3.217 - Forward+ - Using Vulkan Device #0: AMD - AMD Radeon(TM) Graphics
 
WARNING: Blend file import is enabled in the project settings, but no Blender path is configured in the editor settings. Blend files will not be imported.
     at: _editor_init (C:\Users\Administrator\Desktop\ms\godot\modules\gltf\register_types.cpp:63)
ERROR: Condition "!preset.is_valid()" is true. Continuing.
   at: EditorExport::load_config (C:\Users\Administrator\Desktop\ms\godot\editor\export\editor_export.cpp:225)
Exception: Exception 0xc0000005 encountered at address 0x000000: User-mode data execution prevention (DEP) violation at location 0x00000000
Exception: Exception 0xc0000005 encountered at address 0x000000: User-mode data execution prevention (DEP) violation at location 0x00000000
Exception: Exception 0xc0000005 encountered at address 0x000000: User-mode data execution prevention (DEP) violation at location 0x00000000
Exception: Exception 0xc0000005 encountered at address 0x000000: User-mode data execution prevention (DEP) violation at location 0x00000000
Exception: Exception 0xc0000005 encountered at address 0x000000: User-mode data execution prevention (DEP) violation at location 0x00000000
Exception: Exception 0xc0000005 encountered at address 0x000000: User-mode data execution prevention (DEP) violation at location 0x00000000
Exception: Exception 0xc0000005 encountered at address 0x000000: User-mode data execution prevention (DEP) violation at location 0x00000000
Exception: Exception 0xc0000005 encountered at address 0x000000: User-mode data execution prevention (DEP) violation at location 0x00000000`

**Reply by Jimmy2099** [reply]
![image](https://github.com/jpakkane/godot/assets/25509024/0be0f542-761e-4bfe-91b9-b3a22406047a)
![image](https://github.com/jpakkane/godot/assets/25509024/78f5eec4-a10b-467c-a547-b0517c7b8b21)

**Reply by Jimmy2099** [reply]
i have to comment this line to let the code run 
![image](https://github.com/jpakkane/godot/assets/25509024/f48de146-caa6-4263-80a6-ea5df32eefef)
